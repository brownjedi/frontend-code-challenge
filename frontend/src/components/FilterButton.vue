<template>
  <button
    type="button"
    data-cy="filter-button"
    :class="[styles.button, active && styles.active]"
    @click="$emit('click', text, color)"
  >
    {{ text }}
  </button>
</template>

<script lang="ts">
import { computed, defineComponent, useCssModule } from 'vue'

export default defineComponent({
  emits: ['click'],
  props: {
    text: {
      type: String,
      required: true,
    },
    active: {
      type: Boolean,
    },
  },
  setup(props) {
    const styles = useCssModule()
    const btnColor = computed(() => `var(--${props.text.toLowerCase()}-text-color)`)
    const color = computed(() =>
      getComputedStyle(document.documentElement).getPropertyValue(`--${props.text.toLowerCase()}-text-color`)
    )

    return { styles, btnColor, color }
  },
})
</script>

<style lang="scss" module>
.button {
  text-transform: uppercase;
  padding: 0.625rem 1rem;
  border: none;
  cursor: pointer;
  color: v-bind(btnColor);
  transition: background-color 0.5s ease;

  &:hover {
    background-color: v-bind(btnColor);
    color: #fff;
  }

  &.active {
    background-color: v-bind(btnColor);
    color: #fff;
  }
}
</style>
