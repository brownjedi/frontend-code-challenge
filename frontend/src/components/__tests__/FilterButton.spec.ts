import { shallowMount } from '@vue/test-utils'
import FilterButton from '@/components/FilterButton.vue'

describe('components/FilterButton.vue', () => {
  it('has the expected html structure', async () => {
    const wrapper = shallowMount(FilterButton, { props: { text: 'Fire' } })
    expect(wrapper.element).toMatchSnapshot()
  })

  it('has active class when active prop is set to true', async () => {
    const wrapper = shallowMount(FilterButton, { props: { text: 'Fire' } })

    expect(wrapper.find('button').classes()).not.toContain('active')

    await wrapper.setProps({ text: 'Fire', active: true })

    expect(wrapper.find('button').classes()).toContain('active')
  })

  it('emits text and color when clicked', async () => {
    // inputs
    const text = 'Fire'
    const color = '#FF0000'

    // document setup
    document.body.innerHTML = `
      <div>
        <h1>Non Vue app</h1>
        <div id="app"></div>
      </div>
    `
    document.documentElement.style.setProperty(`--${text.toLowerCase()}-text-color`, color)

    const wrapper = shallowMount(FilterButton, { attachTo: '#app', props: { text } })

    await wrapper.find('button').trigger('click')

    const clickEventResponse = wrapper.emitted().click[0] as string[]
    expect(clickEventResponse[0]).toStrictEqual(text)
    expect(clickEventResponse[1]).toStrictEqual(color)

    wrapper.unmount()
  })
})
