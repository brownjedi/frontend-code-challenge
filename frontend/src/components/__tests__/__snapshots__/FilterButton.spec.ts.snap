// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/FilterButton.vue has the expected html structure 1`] = `
<button
  class="button"
  data-cy="filter-button"
  type="button"
>
  Fire
</button>
`;
